File: src/main/java/no/nav/tjenestepensjon/simulering/config/ClaimSetVerifierConfig.java
package no.nav.tjenestepensjon.simulering.config;

import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.oauth2.common.exceptions.InvalidTokenException;
import org.springframework.security.oauth2.provider.token.store.DelegatingJwtClaimsSetVerifier;
import org.springframework.security.oauth2.provider.token.store.JwtClaimsSetVerifier;
import org.springframework.util.Assert;
import org.springframework.util.CollectionUtils;

import no.nav.tjenestepensjon.simulering.config.TokenProviderConfig.TokenProvider;

@Configuration
public class ClaimSetVerifierConfig {

    private final List<TokenProvider> tokenProviders;

    public ClaimSetVerifierConfig(List<TokenProvider> tokenProviders) {
        this.tokenProviders = tokenProviders;
    }

    @Bean
    public JwtClaimsSetVerifier claimsSetVerifier() {
        return new DelegatingJwtClaimsSetVerifier(List.of(new IssuerClaimVerifier(tokenProviders.stream().map(TokenProvider::getIssuer).collect(Collectors.toSet()))));
    }

    static class IssuerClaimVerifier implements JwtClaimsSetVerifier {
        private static final String ISS_CLAIM = "iss";
        private final Set<String> issuerUrls;

        public IssuerClaimVerifier(Set<String> issuerUrls) {
            Assert.notNull(issuerUrls, "issuerUrls cannot be null");
            Assert.notEmpty(issuerUrls, "issuerUrls cannot be empty");
            this.issuerUrls = issuerUrls;
        }

        @Override
        public void verify(Map<String, Object> claims) {
            if (!CollectionUtils.isEmpty(claims) && claims.containsKey(ISS_CLAIM)) {
                String jwtIssuer = (String) claims.get(ISS_CLAIM);
                if (!issuerUrls.contains(jwtIssuer)) {
                    throw new InvalidTokenException("Invalid issuer (iss) claim: " + jwtIssuer);
                }
            }
        }
    }
}

File: src/main/kotlin/no/nav/tjenestepensjon/simulering/config/ClaimSetVerifierConfig.kt
package no.nav.tjenestepensjon.simulering.config

import no.nav.tjenestepensjon.simulering.config.TokenProviderConfig.TokenProvider
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.oauth2.common.exceptions.InvalidTokenException
import org.springframework.security.oauth2.provider.token.store.DelegatingJwtClaimsSetVerifier
import org.springframework.security.oauth2.provider.token.store.JwtClaimsSetVerifier
import org.springframework.util.Assert

@Configuration
class ClaimSetVerifierConfig(private val tokenProviders: List<TokenProvider>) {
    @Bean
    fun claimsSetVerifier() =
            DelegatingJwtClaimsSetVerifier(listOf(IssuerClaimVerifier(tokenProviders.map(TokenProvider::issuer).toSet())))

    internal class IssuerClaimVerifier(private val issuerUrls: Set<String>) : JwtClaimsSetVerifier {

        override fun verify(claims: Map<String, Any>) {
            claims[ISS_CLAIM]?.toString()?.takeUnless(issuerUrls::contains).let {
                throw InvalidTokenException("Invalid issuer (iss) claim: $it")
            }
        }

        companion object {
            private const val ISS_CLAIM = "iss"
        }

        init {
            Assert.notNull(issuerUrls, "issuerUrls cannot be null")
            Assert.notEmpty(issuerUrls, "issuerUrls cannot be empty")
        }
    }

}